File: leetcode/101-200/199-binary-tree-right-side-view/bench/right_side_view.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    long val;
    long left;
    long right;
} Node;

/* Returns the number of visible values and writes them into out[]. */
static long right_view(const Node *nodes, long root, long *out) {
    long count = 0;
    if (root == -1) {
        return 0;
    }
    long capacity = 16;
    long *level = malloc((size_t)capacity * sizeof(long));
    long level_len = 0;
    level[level_len++] = root;
    while (level_len > 0) {
        out[count++] = nodes[level[level_len - 1]].val;
        long *nxt = malloc((size_t)(level_len * 2 + 1) * sizeof(long));
        long nxt_len = 0;
        for (long j = 0; j < level_len; j++) {
            long idx = level[j];
            if (nodes[idx].left != -1) {
                nxt[nxt_len++] = nodes[idx].left;
            }
            if (nodes[idx].right != -1) {
                nxt[nxt_len++] = nodes[idx].right;
            }
        }
        free(level);
        level = nxt;
        level_len = nxt_len;
    }
    free(level);
    return count;
}

int main(void) {
    long n = 8191;
    long passes = 40000;

    Node *nodes = malloc((size_t)n * sizeof(Node));
    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245 + 12345) & 2147483647;
        long v = (state >> 16) % 1000;
        long li = 2 * i + 1;
        long ri = 2 * i + 2;
        nodes[i].val = v;
        nodes[i].left = li < n ? li : -1;
        nodes[i].right = ri < n ? ri : -1;
    }

    long *out = malloc((size_t)n * sizeof(long));
    long sink = 0;
    for (long p = 0; p < passes; p++) {
        state = (state * 1103515245 + 12345) & 2147483647;
        long idx = state % n;
        nodes[idx].val = (state >> 16) % 1000;
        long vn = right_view(nodes, 0, out);
        for (long k = 0; k < vn; k++) {
            sink += out[k];
        }
    }
    printf("%ld\n", sink);
    free(out);
    free(nodes);
    return 0;
}

File: leetcode/101-200/199-binary-tree-right-side-view/bench/right_side_view.rs
struct Node {
    val: i64,
    left: i64,
    right: i64,
}

fn right_view(nodes: &[Node], root: i64) -> Vec<i64> {
    let mut result: Vec<i64> = Vec::new();
    if root == -1 {
        return result;
    }
    let mut level: Vec<i64> = Vec::new();
    level.push(root);
    while !level.is_empty() {
        result.push(nodes[level[level.len() - 1] as usize].val);
        let mut next: Vec<i64> = Vec::new();
        for &idx in &level {
            if nodes[idx as usize].left != -1 {
                next.push(nodes[idx as usize].left);
            }
            if nodes[idx as usize].right != -1 {
                next.push(nodes[idx as usize].right);
            }
        }
        level = next;
    }
    result
}

fn main() {
    let n: i64 = 8191;
    let passes: i64 = 40000;

    let mut nodes: Vec<Node> = Vec::with_capacity(n as usize);
    let mut state: i64 = 12345;
    for i in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        let v = (state >> 16) % 1000;
        let li = 2 * i + 1;
        let ri = 2 * i + 2;
        nodes.push(Node {
            val: v,
            left: if li < n { li } else { -1 },
            right: if ri < n { ri } else { -1 },
        });
    }

    let mut sink: i64 = 0;
    for _ in 0..passes {
        state = (state * 1103515245 + 12345) & 2147483647;
        let idx = (state % n) as usize;
        nodes[idx].val = (state >> 16) % 1000;
        let view = right_view(&nodes, 0);
        for v in &view {
            sink += *v;
        }
    }
    println!("{}", sink);
}
